pkg/catalogv2/git/download.go

```diff
@@ -1,124 +1,125 @@
 package git
 
 import (
-	"crypto/sha256"
-	"encoding/hex"
-	"encoding/pem"
 	"fmt"
-	"net/url"
-	"os"
-	"path/filepath"
-	"regexp"
-	"strings"
 
 	"github.com/rancher/rancher/pkg/settings"
 	corev1 "k8s.io/api/core/v1"
 )
 
-const (
-	stateDir  = "management-state/git-repo"
-	staticDir = "/var/lib/rancher-data/local-catalogs/v2"
-	localDir  = "../rancher-data/local-catalogs/v2" // identical to helm.InternalCatalog
-)
+// Ensure builds the configuration for a should-existing repo and makes sure it is cloned or reseted to the latest commit of given branch
+func Ensure(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) error {
```

Comment:
I made this structure of separated methods to achieve more maintainability. 
All these methods need to call **gitForRepo**(.....) with a bunch of variables before and then call the real action method later sometimes with branch and sometimes with commit and etc... 

I know it is less beautiful but I regard it as more maintainable since it is easier to see what is going in and out from the function. I can merge **Ensure + EnsureClonedRepo** and **Clone + CloneToHead** however, I can't do it with **Update** + **UpdateToLatestRef** since **Update** has some logic inside. 

Also in the future depending on the necessity to change **Ensure or Clone** behavior, it would be easier if it is decoupled like I did.

Reply:
The gitForRepo() is creating a new Git client struct which has a bunch of functions that are later called. 

The same pattern is followed for Head and Update. 

why can't we merged Update + UpdatetoLatestRef ?